File: articles/rise-of-modern-biological-thought/python/historical_timeline.py
"""
Historical timeline summary workflow.

Run:
    python python/historical_timeline.py
"""

from __future__ import annotations

from pathlib import Path

import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
TIMELINE_PATH = ARTICLE_DIR / "data" / "historical_milestones.csv"


def main() -> None:
    timeline = pd.read_csv(TIMELINE_PATH)

    timeline["century"] = ((timeline["year"] - 1) // 100 + 1).astype(int)

    domain_summary = (
        timeline.groupby("domain")
        .agg(
            first_year=("year", "min"),
            last_year=("year", "max"),
            n_milestones=("milestone", "count"),
        )
        .reset_index()
        .sort_values(["first_year", "domain"])
    )

    century_summary = (
        timeline.groupby("century")
        .agg(
            first_year=("year", "min"),
            last_year=("year", "max"),
            n_milestones=("milestone", "count"),
            domains=("domain", lambda x: ", ".join(sorted(set(x)))),
        )
        .reset_index()
        .sort_values("century")
    )

    print(domain_summary.to_string(index=False))
    print(century_summary.to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/rise-of-modern-biological-thought/r/historical_timeline.R
# Historical timeline summary in R.

timeline_path <- file.path("data", "historical_milestones.csv")
if (!file.exists(timeline_path)) {
  timeline_path <- file.path("..", "data", "historical_milestones.csv")
}

timeline <- read.csv(timeline_path)

timeline$century <- floor((timeline$year - 1) / 100) + 1

domain_summary <- aggregate(
  year ~ domain,
  data = timeline,
  FUN = function(x) c(first_year = min(x), last_year = max(x), n = length(x))
)

print(timeline[order(timeline$year), ])
print(domain_summary)
